clang/runtime/dpct-rt/include/dpct/blas_utils.hpp

```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
+    from_linear_idx = a_ld * col_idx + row_idx;
+
+    size_t from_row_in_row32_col32 = row_idx % 32;
+    size_t from_col_in_row32_col32 = col_idx % 32;
+
+    size_t to_row_in_row32_col32 = 8 * ((from_row_in_row32_col32 % 8) / 2) +
+                                   (from_row_in_row32_col32 / 8) * 2 +
+                                   from_row_in_row32_col32 % 2;
+    size_t to_col_in_row32_col32 = from_col_in_row32_col32;
+    size_t to_linear_idx_in_row32_col32 =
+        to_row_in_row32_col32 * 32 + to_col_in_row32_col32;
+
+    to_linear_idx = c_ld * (col_idx / 32) + (row_idx / 32) * (32 * 32) +
+                    to_linear_idx_in_row32_col32;
+    break;
+  }
+  case COMBINE(order_t::col32_2r_4r4, order_t::col): {
+    to_linear_idx = c_ld * col_idx + row_idx;
+
+    size_t to_row_in_row32_col32 = row_idx % 32;
+    size_t to_col_in_row32_col32 = col_idx % 32;
+
+    size_t from_row_in_row32_col32 = 8 * ((to_row_in_row32_col32 % 8) / 2) +
+                                     (to_row_in_row32_col32 / 8) * 2 +
+                                     to_row_in_row32_col32 % 2;
+    size_t from_col_in_row32_col32 = to_col_in_row32_col32;
+    size_t from_linear_idx_in_row32_col32 =
+        from_row_in_row32_col32 * 32 + from_col_in_row32_col32;
+
+    from_linear_idx = a_ld * (col_idx / 32) + (row_idx / 32) * (32 * 32) +
+                      from_linear_idx_in_row32_col32;
+    break;
+  }
+  }
+#undef COMBINE
+  return std::make_tuple(from_linear_idx, to_linear_idx);
+}
+
+template <class T>
+sycl::event matrix_transform(queue_ptr q_ptr, size_t rows, size_t cols,
+                             size_t a_ld, order_t a_order, const T *a,
+                             size_t c_ld, order_t c_order, T *c) {
+  if ((a_order != order_t::col && c_order != order_t::col) ||
+      (a_order == order_t::col && c_order == order_t::col)) {
+    throw std::runtime_error("Unsupported order combination.");
+  }
+
+  return q_ptr->submit([&](sycl::handler &cgh) {
+#ifdef DPCT_USM_LEVEL_NONE
+    access_wrapper<const T *> a_acc(a, cgh);
+    access_wrapper<T *> c_acc(c, cgh);
+#endif
+    cgh.parallel_for<dpct_kernel_name<class matrix_transform_col_to_row, T>>(
+        sycl::range<2>(a_ld, cols), [=](sycl::id<2> index) {
+#ifdef DPCT_USM_LEVEL_NONE
+          auto a_data = a_acc.get_raw_pointer();
+          auto c_data = c_acc.get_raw_pointer();
+#else
+            auto a_data = a;
+            auto c_data = c;
+#endif
+          size_t row_idx = index.get(0);
+          size_t col_idx = index.get(1);
+          if (row_idx < rows) {
+            size_t from_linear_idx, to_linear_idx;
+            std::tie(from_linear_idx, to_linear_idx) = get_from_to_linear_idx(
+                rows, cols, a_ld, a_order, c_ld, c_order, row_idx, col_idx);
+            c_data[to_linear_idx] = a_data[from_linear_idx];
+          }
+        });
+  });
+}
+} // namespace detail
+
+inline void matmul(lt_handle_ptr handle, matmul_desc_ptr compute_desc,
```

Comment:
this API is async api, how about return the event which end user can wait on.

Reply:
done